qiskit/quantum_info/operators/symplectic/clifford.py

```diff
@@ -424,72 +431,99 @@ def compose(self, other, qargs=None, front=False):
-
-        for k in range(2 * num_qubits):
+        left, right = (self, other) if front else (other, self)
 
-            row2 = array2[k]
-            x2 = table2.x[k]
-            z2 = table2.z[k]
-
-            # Adding a factor of i for each Y in the image of an operator under the
-            # first operation, since Y=iXZ
-
-            ifacts[k] += np.sum(x2 & z2)
-
-            # Adding factors of i due to qubit-wise Pauli multiplication
-
-            for j in range(num_qubits):
-                x = 0
-                z = 0
-                for i in range(2 * num_qubits):
-                    if row2[i]:
-                        x1 = array1[i, j]
-                        z1 = array1[i, j + num_qubits]
-                        if (x | z) & (x1 | z1):
-                            val = np.mod(np.abs(3 * z1 - x1) - np.abs(3 * z - x) - 1, 3)
-                            if val == 0:
-                                ifacts[k] += 1
-                            elif val == 1:
-                                ifacts[k] -= 1
-                        x = np.mod(x + x1, 2)
-                        z = np.mod(z + z1, 2)
+        if self.num_qubits == 1:
+            return self._compose_1q(left, right)
+        return self._compose_general(left, right)
 
+    @classmethod
+    def _compose_general(cls, first, second):
+        # Correcting for phase due to Pauli multiplication. Start with factors of -i from XZ = -iY
+        # on individual qubits, and then handle multiplication between each qubitwise pair.
+        ifacts = np.sum(second.x & second.z, axis=1, dtype=int)
+
+        # A lookup table for calculating phases.  The indices are
+        #     current_x, current_z, running_x_count, running_z_count
+        # where all counts taken modulo 2.
+        lookup = np.zeros((2, 2, 2, 2), dtype=int)
+        lookup[0, 1, 1, 0] = lookup[1, 0, 1, 1] = lookup[1, 1, 0, 1] = -1
+        lookup[0, 1, 1, 1] = lookup[1, 0, 0, 1] = lookup[1, 1, 1, 0] = 1
+
+        x1, z1 = first.x.astype(np.uint8), first.z.astype(np.uint8)
+        lookup = cls._compose_lookup()
+
+        # The loop is over 2*n_qubits entries, and the entire loop is cubic in the number of qubits.
+        for k, row2 in enumerate(second.symplectic_matrix):
+            x1_select = x1[row2]
+            z1_select = z1[row2]
+            x1_accum = np.logical_xor.accumulate(x1_select, axis=0).astype(np.uint8)
+            z1_accum = np.logical_xor.accumulate(z1_select, axis=0).astype(np.uint8)
+            indexer = (x1_select[1:], z1_select[1:], x1_accum[:-1], z1_accum[:-1])
+            ifacts[k] += np.sum(lookup[indexer])
         p = np.mod(ifacts, 4) // 2
 
-        phase = np.mod(phase + p, 2)
+        phase = (
+            (np.matmul(second.symplectic_matrix, first.phase, dtype=int) + second.phase + p) % 2
+        ).astype(bool)
+        data = cls._stack_table_phase(
+            (np.matmul(second.symplectic_matrix, first.symplectic_matrix, dtype=int) % 2).astype(
```

Comment:
iirc correctly it was meaningfully slower, but it was a long time since I wrote this.

Reply:
Oh interesting, I guess it depends pretty heavily on the numpy implementations. In my current environment, the `@` seems to be winning
![image](https://user-images.githubusercontent.com/2229105/205322862-2e106b15-055d-4f16-824a-8ac0f9102f79.png)

Reply:
Oh no, it's because they're not the same thing.  Numpy implements matrix multiplication over bools using the natural `+`, which in bool logic is more akin to `logical_or`.  I can't think of a way off the top of my head to get Numpy to use a generalised version using `xor` as the reduction instead of `or` (without manually allocating a massive matrix as an intermediary), but maybe there is?

Reply:
If this is the bottleneck, I guess we can pull it into Rust to sort it out faster, but I think in practice the scaling of the factors of :math:`i` is worse, so it might not matter too much.

Reply:
Oh you're right, I forgot to put the vital `%2` in there